# ~~First~~ Fast Python Notebook

<div style="max-width: 640px">

An accelerated guide to analyzing data with the <a href="https://www.python.org/">Python</a> programming language and a <a href="https://jupyter.org/">Jupyter</a> notebook

</div>
    
By [Ben Welsh](https://palewi.re/who-is-ben-welsh/)

<div style="max-width: 640px">

First developed in 2016, <a href="https://palewi.re/docs/first-python-notebook/">"First Python Notebook"</a> is a tutorial that guides students through a data-driven investigation of money in California politics. It is most commonly taught as a six-hour, in-person class. This document is an abbreviated spinoff intended to be taught online in two to three hours.

You will learn just enough of the Python computer programming language to work with the <a href="https://pandas.pydata.org/">pandas</a> library, a popular open-source tool for analyzing data. The course will teach you how to read, filter, join, group, aggregate and rank structured data by recreating a helicopter accident analysis <a href="https://github.com/datadesk/helicopter-accident-analysis">published by the Los Angeles Times</a>.
    
</div>

## What is a Jupyter notebook?

<div style="max-width: 640px">

<p><img src="https://palewi.re/docs/first-python-notebook/_static/img/labpreview.webp" alt="jupyter"></p>
<p>A <a href="https://jupyter.org/">Jupyter</a> notebook is a browser-based interface where you can write, run, remix and republish code. It is free software that anyone can install and run.</p>
<p><a href="https://nbviewer.jupyter.org/github/robertodealmeida/notebooks/blob/master/earth_day_data_challenge/Analyzing%20whale%20tracks.ipynb">Scientists</a>, <a href="https://nbviewer.jupyter.org/github/nealcaren/workshop_2014/blob/master/notebooks/5_Times_API.ipynb">scholars</a>, <a href="https://github.com/rsvp/fecon235/blob/master/nb/fred-debt-pop.ipynb">investors</a> and <a href="https://netflixtechblog.com/notebook-innovation-591ee3221233">corporations</a> use Jupyter to create and share their research. It is also used by journalists to develop stories and show their work. Examples include:</p>
<ul>
<li><a href="https://github.com/BuzzFeedNews/2016-01-tennis-betting-analysis/blob/master/notebooks/tennis-analysis.ipynb">“The Tennis Racket”</a> by BuzzFeed and the BBC</li>
<li><a href="https://github.com/propublica/compas-analysis/blob/master/Compas%20Analysis.ipynb">“Machine bias”</a> by ProPublica</li>
<li><a href="https://github.com/asuozzo/arcos-opioid-analysis-vt">“As Opioid Crisis Ramped Up, Pills Flowed Into Vermont by the Millions”</a> by Seven Days</li>
<li><a href="https://github.com/datadesk/notebooks">More than 35 different notebooks</a> published by the Los Angeles Times</li>
</ul>
<p>There are numerous ways to install and configure Jupyter notebooks. This class is taught using <a href="https://jupyterlite.readthedocs.io/">JupyterLite</a> a lightweight distribution that runs entirely in your web browser. For instructions on how to install a more powerful version on your computer consult <a href="https://palewi.re/docs/first-python-notebook/jupyter_desktop.html">the full edition of &quot;First Python Notebook.&#39;</a></p>
<p>Once you have this notebook up and running, you&#39;re ready to write Python in a code cell. Do not stress. There is nothing too fancy about it. You can start by just doing a little simple math.</p>
<p>Select the box below, then hit the play button in the toolbar above the notebook or hit <code>SHIFT+ENTER</code> on your keyboard.</p>
    
</div>

In [1]:
2+2

4

<div style="max-width:630px">

There. You have just run your first Python code. You have entered two integers and added them together using the plus sign operator.

Not so bad, right? Now try writing in your own math problem in the next cell. Maybe `2+3` or `2+200`. Whatever strikes your fancy. After you've typed it in, hit the play button or `SHIFT+ENTER`.

</div>

<div style="max-width: 640px;">

This to-and-fro of writing Python code in a cell and then running it is the rhythm of working in a notebook. If you get an error after you run a cell, look carefully at your code and see that it exactly matches what’s been written in the example.
    
Here's an example of a error that I've added intentionally:
    
</div>

In [4]:
2+2+

SyntaxError: invalid syntax (4150814810.py, line 1)

<div style="max-width: 640px;">

Don’t worry. Code crashes are a normal part of life for computer programmers. They’re usually caused by small typos that can be quickly corrected.
    
</div>

In [3]:
2+2+2

6

<div style="max-width: 640px;">

<p>Over time you will gradually stack cells to organize an analysis that runs from top to bottom. The cells can contain variables, functions and other Python tools.</p>
<div class="alert alert-block alert-warning">
<p>If you’ve never written code before, we recommend <a href="https://docs.python.org/3/tutorial/introduction.html">&quot;An Informal Introduction to Python&quot;</a> and subsequent sections of python.org’s tutorial.</p>
</div>
<p>A simple example would be storing your number in a variable in one cell:</p>

    
</div>

In [4]:
number = 2

Then adding it to another number in the next:

In [5]:
number + 3

5

Change the `number` value to 3 and run both cells again. Instead of 5, it should now output 6.

In [5]:
number = 3

In [6]:
number + 3

6

Now try defining your own numeric variable and doing some math with it. You can name it whatever you want. Want to try some other math operations? The `-` sign does subtraction. Multipication is `*`. Division is `/`.

Once you’ve got the hang of making the notebook run, you’re ready to introduce pandas, a powerful Python analysis library that can do a whole lot more than add a few numbers together.

## What is pandas? 

<div style="max-width: 640px">
  
<p><img src="https://palewi.re/docs/first-python-notebook/_static/img/pandas-pypi.png" alt="pandas on the Python Package Index"></p>
<p>Lucky for us, Python is filled with functions to do almost anything you’d want to do with a programming language: <a href="http://docs.python-requests.org/">navigate the web</a>, <a href="https://docs.python.org/2/library/csv.html">parse data</a>, <a href="http://www.sqlalchemy.org/">interact with a database</a>, <a href="https://www.scipy.org/">run fancy statistics</a>, <a href="https://www.djangoproject.com/">build a pretty website</a> and <a href="https://www.crummy.com/software/BeautifulSoup/">so</a> <a href="http://www.nltk.org/">much</a> <a href="https://pillow.readthedocs.io/en/stable/">more</a>.</p>
<p>Creative people have put these tools to work to get a <a href="https://www.python.org/about/success/">wide range of things</a> done in the academy, the laboratory and even in outer space.</p>
<p>Some of those tools are included in a toolbox that comes with the language, known as the standard library. Others have been built by members of Python’s developer community and need to be separately downloaded and installed. One third-party tool that’s important for this class is called <a href="https://pandas.pydata.org/">pandas</a>. Invented by programmers at a <a href="https://www.aqr.com/">financial investment firm</a>, it has become a leading open-source library for accessing and analyzing data.</p>
<p>Here’s how to use pandas yourself. Run the following:</p>
    
</div>

In [8]:
import pandas

<div style="max-width: 640px">

<p>If nothing happens, that’s good. It means you have it installed and ready as to use.</p>
<div class="alert alert-block alert-warning">
<p>Since pandas is created by a third party independent from the core Python developers, it may not be available by default if you manually installed Python and Jupyter. It’s available here because JupyterLite, whose developers have curated a list of common utilities to include with their distribution. Consult our <a href="https://palewi.re/docs/first-python-notebook/appendix/index.html">advanced installation guide</a> if the cell above threw an error.</p>
</div>
<p>Now let&#39;s run the same code again, but with a small addition.</p>

    
</div>

In [10]:
import pandas as pd

<div style="max-width: 640px">

This will alias the pandas library at the shorter variable name of `pd`. This is standard practice in the pandas community. You will frequently see examples of pandas code online using pd as shorthand. It’s not required, but it’s good to get in the habit so that your code will be better understood by other computer programmers.

Those two little letters contain dozens of data analysis tools that we’ll use in future lessons. They can import massive data files, compute advanced statistics, filter, sort, rank and do just about anything else you’d want to do.

We’ll get to all of that soon enough, but let’s start out with something simple. Let's run some simple stats.

## Calculating descriptive statistics

Start by making a list of numbers in a new notebook cell. To keep things simple, we'll start with all of the even numbers between zero and ten. Note the variable name I've assigned. Then press play.
    
</div>

In [15]:
my_list = [2, 4, 6, 8]

<div style="max-width: 640px">

If you’re a skilled Python programmer, you can do some cool stuff with any list, including run statistics. But if you hand over to pandas instead, you’ll be impressed by how easily you can analyze the data without much computer code.


In this case, it’s as simple as converting that plain Python list into what pandas calls a [`Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html). Here’s how to make it happen:
    
</div>

In [16]:
my_series = pd.Series(my_list)

<div style="max-width: 640px">

Once the data becomes a `Series`, you can immediately run a wide range of <a href="https://en.wikipedia.org/wiki/Descriptive_statistics">descriptive statistics</a>. Let’s try a few.

First, let’s sum all the numbers.
    
</div>

In [17]:
my_series.sum()

20

Then find the maximum value.

In [18]:
my_series.max()

8

The minimum value.

In [19]:
my_series.min()

2

How about the average, which also known as the mean?

In [20]:
my_series.mean()

5.0

The median?

In [21]:
my_series.median()

5.0

The standard deviation?

In [22]:
my_series.std()

2.581988897471611

Finally, all of the above, plus a little more about the distribution, in one simple command.

In [18]:
my_series.describe()

count    4.000000
mean     5.000000
std      2.581989
min      2.000000
25%      3.500000
50%      5.000000
75%      6.500000
max      8.000000
dtype: float64

<div style="max-width: 640px">

Before you move on, go back the `my_list` variable and change the list. Maybe add a few more values. Or switch to odds. Then rerun all the cells above. You'll see all the statistics update to reflect the different dataset.

Substitute in a series of 10 million records and your notebook would calculate all the same statistics without you needing to write any more code. Once your data, however large or complex, is imported into pandas, simple statistics become a snap.

</div>

<div style="max-width:640px">

## Introducing DataFrames
    
Now it’s time to get our hands on some real data. In 2018, the Los Angeles Times published an investigation headlined, <a href="https://www.latimes.com/projects/la-me-robinson-helicopters/">"The Robinson R44, the world’s best-selling civilian helicopter, has a long history of deadly crashes".</a>
    
It reported that the Robinson R44 led all major models with the highest fatal accident rate from 2006 to 2016. The analysis was <a href="https://github.com/datadesk/helicopter-accident-analysis">published on GitHub</a> as a series of Jupyter notebooks. 

The analysis was based on two key datasets: 
   
1. The National Transportation Safety Board's <a href="https://www.ntsb.gov/_layouts/ntsb.aviation/index.aspx">Aviation Accident Database</a>
2. The Federal Aviation Administration's <a href="https://www.faa.gov/data_research/aviation_data_statistics/general_aviation/">General Aviation and Part 135 Activity Survey</a>

After a significant amount of work gathering and cleaning the source data, the number of accidents for each helicopter model were normalized using the flight hours estimates in the survey. For the purposes of this demonstration, we will read in tidied versions of each file that are ready for analysis.
    
The data are structured in rows of comma-separated values. This is known as a [CSV file](https://en.wikipedia.org/wiki/Comma-separated_values). It is the most common way you will find data published online.

The pandas library is able to read in files from a variety formats, including CSV. In our next cell, we'll use pandas' `read_csv` method to read in `ntsb-accidents.csv`.
    
</div>

In [29]:
pd.read_csv("ntsb-accidents.csv")

event_id  aircraft_id                    ntsb_make ntsb_model  \
0    20121010X41638            1                         BELL       206B   
1    20130222X12624            1                   EUROCOPTER  AS 350 B2   
2    20140127X44720            1                         BELL     206L 3   
3    20091012X15520            1    MCDONNELL DOUGLAS HELI CO      369FF   
4    20161116X00108            1                       HUGHES  369D/500D   
..              ...          ...                          ...        ...   
158  20070802X01089            1                   EUROCOPTER   AS350-B2   
159  20070802X01089            2                   EUROCOPTER   AS350-B2   
160  20060605X00682            1                         BELL     206L-3   
161  20140717X70001            1                   AGUSTA SPA      A109E   
162  20160612X85856            1  ROBINSON HELICOPTER COMPANY     R44 II   

    ntsb_number  year               date             city state country  \
0    CEN13FA009  2012  10/10/12 00:00:00          CROWLEY    LA     USA   
1    CEN13FA174  2013  02/22/13 00:00:00    OKLAHOMA CITY    OK     USA   
2    CEN14FA122  2014  01/27/14 00:00:00             SILT    CO     USA   
3    WPR10FA012  2009  10/12/09 00:00:00           FOSSIL    OR     USA   
4    WPR17FA021  2016  11/15/16 00:00:00            PUKOO    HI     USA   
..          ...   ...                ...              ...   ...     ...   
158  LAX07MA231  2007  07/27/07 00:00:00          PHOENIX    AZ     USA   
159  LAX07MA231  2007  07/27/07 00:00:00          PHOENIX    AZ     USA   
160  DFW06IA145  2006  05/30/06 00:00:00  GRAND ISLE 43AA    GM     USA   
161  CEN14FA369  2014  07/17/14 00:00:00          NEWKIRK    NM     USA   
162  CEN16FA215  2016  06/12/16 00:00:00        JONESBORO    AR     USA   

     total_fatalities  in_usa  pilot_total       latimes_make latimes_model  \
0                 1.0    True          1.0               BELL           206   
1                 2.0    True          1.0             AIRBUS           350   
2                 3.0    True          1.0               BELL           206   
3                 1.0    True          1.0  MCDONNELL DOUGLAS           369   
4                 2.0    True          1.0             HUGHES           369   
..                ...     ...          ...                ...           ...   
158               4.0    True          1.0             AIRBUS           350   
159               4.0    True          1.0             AIRBUS           350   
160               1.0    True          1.0               BELL           206   
161               3.0    True          1.0             AGUSTA           109   
162               1.0    True          1.0           ROBINSON           R44   

     count latimes_make_and_model  
0    112.0               BELL 206  
1      8.0             AIRBUS 350  
2      3.0               BELL 206  
3      3.0  MCDONNELL DOUGLAS 369  
4      1.0             HUGHES 369  
..     ...                    ...  
158    3.0             AIRBUS 350  
159    3.0             AIRBUS 350  
160   20.0               BELL 206  
161    2.0             AGUSTA 109  
162   22.0           ROBINSON R44  

[163 rows x 17 columns]

<div style="max-width:640px">

You should see a big table like the one above. It is a DataFrame where pandas has structured the CSV data into rows and columns, just like Excel or other spreadsheet software might.

A major advantage of Jupyter over spreadsheets is that rather than manipulating the data through a haphazard series of clicks and keypunches we will be gradually grinding it down using a computer programming script that is transparent and reproducible.

In order to do more with your DataFrame, we need to store it so it can be reused in subsequent cells. We can do this by saving in a variable, which is a fancy computer programming word for a named shortcut where we save our work as we go.

</div>

In [30]:
accident_list = pd.read_csv("ntsb-accidents.csv")

<div style="max-width:640px">

After you run it, you shouldn’t see anything. That’s a good thing. It means our DataFrame has been saved under the name `accident_list`, which we can now begin interacting with in the cells that follow.

We can do this by calling “methods” that pandas makes available to all DataFrames. You may not have known it at the time, but `read_csv` is one of these methods. There are dozens more that can do all sorts of interesting things. Let’s start with some easy ones that analysts use all the time.

To preview the first few rows of the dataset, try the <a href="https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html">`head`</a> method.
    
</div>

In [31]:
accident_list.head()

event_id  aircraft_id                  ntsb_make ntsb_model  \
0  20121010X41638            1                       BELL       206B   
1  20130222X12624            1                 EUROCOPTER  AS 350 B2   
2  20140127X44720            1                       BELL     206L 3   
3  20091012X15520            1  MCDONNELL DOUGLAS HELI CO      369FF   
4  20161116X00108            1                     HUGHES  369D/500D   

  ntsb_number  year               date           city state country  \
0  CEN13FA009  2012  10/10/12 00:00:00        CROWLEY    LA     USA   
1  CEN13FA174  2013  02/22/13 00:00:00  OKLAHOMA CITY    OK     USA   
2  CEN14FA122  2014  01/27/14 00:00:00           SILT    CO     USA   
3  WPR10FA012  2009  10/12/09 00:00:00         FOSSIL    OR     USA   
4  WPR17FA021  2016  11/15/16 00:00:00          PUKOO    HI     USA   

   total_fatalities  in_usa  pilot_total       latimes_make latimes_model  \
0               1.0    True          1.0               BELL           206   
1               2.0    True          1.0             AIRBUS           350   
2               3.0    True          1.0               BELL           206   
3               1.0    True          1.0  MCDONNELL DOUGLAS           369   
4               2.0    True          1.0             HUGHES           369   

   count latimes_make_and_model  
0  112.0               BELL 206  
1    8.0             AIRBUS 350  
2    3.0               BELL 206  
3    3.0  MCDONNELL DOUGLAS 369  
4    1.0             HUGHES 369

It does the first five by default. If you want a different number, submit it as an input.

In [32]:
accident_list.head(1)

event_id  aircraft_id ntsb_make ntsb_model ntsb_number  year  \
0  20121010X41638            1      BELL       206B  CEN13FA009  2012   

                date     city state country  total_fatalities  in_usa  \
0  10/10/12 00:00:00  CROWLEY    LA     USA               1.0    True   

   pilot_total latimes_make latimes_model  count latimes_make_and_model  
0          1.0         BELL           206  112.0               BELL 206

To get a look at all of the columns and what type of data they store, try the [`info`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.info.html) method.

In [33]:
accident_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 163 entries, 0 to 162
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   event_id                163 non-null    object 
 1   aircraft_id             163 non-null    int64  
 2   ntsb_make               163 non-null    object 
 3   ntsb_model              163 non-null    object 
 4   ntsb_number             163 non-null    object 
 5   year                    163 non-null    int64  
 6   date                    163 non-null    object 
 7   city                    163 non-null    object 
 8   state                   162 non-null    object 
 9   country                 163 non-null    object 
 10  total_fatalities        163 non-null    float64
 11  in_usa                  163 non-null    bool   
 12  pilot_total             163 non-null    float64
 13  latimes_make            163 non-null    object 
 14  latimes_model           163 non-null    ob

Look carefully at the results and you'll see we have 163 fatal accidents.

## Inspecting columns

<div style="max-width: 640px;">

To see the contents of a column separate from the rest of the DataFrame, add the column’s name to the DataFrame’s variable following a period. We’ll begin with the `latimes_make_and_model` column, which records the standardized name of the helicopter that crashed.

</div>

In [36]:
accident_list.latimes_make_and_model

0                   BELL 206
1                 AIRBUS 350
2                   BELL 206
3      MCDONNELL DOUGLAS 369
4                 HUGHES 369
               ...          
158               AIRBUS 350
159               AIRBUS 350
160                 BELL 206
161               AGUSTA 109
162             ROBINSON R44
Name: latimes_make_and_model, Length: 163, dtype: object

<div style="max-width: 640px;">

That will list the column out as a `Series`, just like the ones we created from scratch earlier. Just as we did then, you can now start tacking on additional methods that will analyze the contents of the column.


<div class="alert alert-block alert-warning">    
    <p>You can also access columns a second way, like this: accident_list['latimes_make_and_model'].</p><p>This method isn’t as pretty, but it’s required if your column has a space in its name, which would break the simpler dot-based method.</p>
</div>
    
In this case, the column is filled with characters. So we don’t want to calculate statistics like the median and average, as we did before.

There’s another built-in pandas tool that will total up the frequency of values in a column. The method is called `value_counts` and it’s just as easy to use as sum, min or max. All you need to do it is add a period after the column name and chain it on the tail end of your cell.

Run the code and you should see the locations ranked by their number of sites.

</div>

In [37]:
accident_list.latimes_make_and_model.value_counts()

ROBINSON R44             38
BELL 206                 30
AIRBUS 350               29
ROBINSON R22             20
HUGHES 369               13
BELL 407                 13
MCDONNELL DOUGLAS 369     6
SCHWEIZER 269             5
AIRBUS 135                4
SIKORSKY 76               2
AGUSTA 109                2
AIRBUS 130                1
Name: latimes_make_and_model, dtype: int64

<div style="max-width: 640px;">

Congratulations, you've made your first finding. With that little line of code, you've calculated an important fact: During the period being studied, the Robinson R44 had more fatal accidents than any other helicopter.
    
You may notice that even though the result has two columns, pandas did not return a clean-looking table in the same way as `head` did for our DataFrame. That’s because our column, a Series, acts a little bit different than the DataFrame created by `read_csv`.

In most instances, if you have an ugly Series generated by a method like `value_counts` and you want to convert it into a pretty DataFram,e you can do so by tacking on the `reset_index` method on the end.

</div>

In [39]:
accident_list.latimes_make_and_model.value_counts().reset_index()

index  latimes_make_and_model
0            ROBINSON R44                      38
1                BELL 206                      30
2              AIRBUS 350                      29
3            ROBINSON R22                      20
4              HUGHES 369                      13
5                BELL 407                      13
6   MCDONNELL DOUGLAS 369                       6
7           SCHWEIZER 269                       5
8              AIRBUS 135                       4
9             SIKORSKY 76                       2
10             AGUSTA 109                       2
11             AIRBUS 130                       1

<div style="max-width: 640px;">

Why does `Series` behave differently than a dataframe? Why does `reset_index` have such a weird name?

Like so much in computer programming, the answer is simply, “because the people who created the library said so.” It’s important to learn that all open-source programming tools are made by humans, and humans have their quirks. Over time you’ll see pandas has more than a few.

As a beginner, you should just accept the oddities and keep moving. As you get more advanced, if there’s something about the system you think could be improved you should consider <a href="https://pandas.pydata.org/pandas-docs/stable/development/contributing.html">contributing</a> to the Python code that operates the library.
    
Before we move on to the next chapter, here's a challenge. See if you can answer a few more questions a journalist might ask about our dataset. All four of the questions below can be answered using only tricks we've covered thus far. See if you can do it.
    
</div>

1. What was the total number of fatalities?

2. Which helicopter maker had the most accidents?

3. What was the total number of helicopter accidents by year?

4. What state had the most helicopter accidents?

## Filtering down the dataset

<div style="max-width: 640px;">

The most common way to filter a DataFrame is to pass an expression as an “index” that can be used to decide which records should be kept and which discarded. You write the expression by combining a column on your DataFrame with an <a href="https://en.wikipedia.org/wiki/Operator_(computer_programming)">“operator”</a> like == or > or < and a value to compare against each row.

<div class="alert alert-block alert-warning">
    <p>If you are familiar with writing <a href="https://en.wikipedia.org/wiki/SQL">SQL</a> to manipulate databases, pandas’ filtering system is somewhat similar to a WHERE query. The <a href="https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sql.html#where">official pandas documentation</a> offers direct translations between the two.</p>
</div>
    
Let's try filtering against the `state` field. Save one of the values listed above into a variable. This will allow us to reuse it later.

</div>

In [41]:
my_state = "IA"

<div style="max-width: 640px;">

In the next cell we will ask pandas to narrow down our list of sites to just those that list the location we’re interested in. We will create a filter expression and place it between two flat brackets following the DataFrame we wish to filter.
 
</div>

In [43]:
accident_list[accident_list.state == my_state]

event_id  aircraft_id                    ntsb_make ntsb_model  \
66   20130102X35708            1              BELL HELICOPTER        407   
117  20060705X00868            1                         BELL       206B   
144  20140908X10448            1  ROBINSON HELICOPTER COMPANY     R44 II   

    ntsb_number  year               date        city state country  \
66   CEN13FA122  2013  01/02/13 00:00:00  CLEAR LAKE    IA     USA   
117  CHI06FA173  2006  06/30/06 00:00:00     WALFORD    IA     USA   
144  CEN14LA487  2014  09/06/14 00:00:00   MACEDONIA    IA     USA   

     total_fatalities  in_usa  pilot_total latimes_make latimes_model  count  \
66                3.0    True          1.0         BELL           407    3.0   
117               1.0    True          1.0         BELL           206  112.0   
144               1.0    True          1.0     ROBINSON           R44   22.0   

    latimes_make_and_model  
66                BELL 407  
117               BELL 206  
144           ROBINSON R44

<div style="max-width: 640px;">

Now we should save the results of that filter into a new variable separate from the full list we imported from the CSV file. Since it includes only the sites for the location we’re interested in let’s call it `my_accidents`.

</div>

In [46]:
my_accidents = accident_list[accident_list.state == my_state]

<div style="max-width: 640px;">

To check our work and find out how many committees are left after the filter, let’s run the DataFrame inspection commands we learned earlier.

First `head`.

</div>

In [47]:
my_accidents.head()

event_id  aircraft_id                    ntsb_make ntsb_model  \
66   20130102X35708            1              BELL HELICOPTER        407   
117  20060705X00868            1                         BELL       206B   
144  20140908X10448            1  ROBINSON HELICOPTER COMPANY     R44 II   

    ntsb_number  year               date        city state country  \
66   CEN13FA122  2013  01/02/13 00:00:00  CLEAR LAKE    IA     USA   
117  CHI06FA173  2006  06/30/06 00:00:00     WALFORD    IA     USA   
144  CEN14LA487  2014  09/06/14 00:00:00   MACEDONIA    IA     USA   

     total_fatalities  in_usa  pilot_total latimes_make latimes_model  count  \
66                3.0    True          1.0         BELL           407    3.0   
117               1.0    True          1.0         BELL           206  112.0   
144               1.0    True          1.0     ROBINSON           R44   22.0   

    latimes_make_and_model  
66                BELL 407  
117               BELL 206  
144           ROBINSON R44

Then `info`.

In [48]:
my_accidents.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3 entries, 66 to 144
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   event_id                3 non-null      object 
 1   aircraft_id             3 non-null      int64  
 2   ntsb_make               3 non-null      object 
 3   ntsb_model              3 non-null      object 
 4   ntsb_number             3 non-null      object 
 5   year                    3 non-null      int64  
 6   date                    3 non-null      object 
 7   city                    3 non-null      object 
 8   state                   3 non-null      object 
 9   country                 3 non-null      object 
 10  total_fatalities        3 non-null      float64
 11  in_usa                  3 non-null      bool   
 12  pilot_total             3 non-null      float64
 13  latimes_make            3 non-null      object 
 14  latimes_model           3 non-null      obj

## Pivoting with `groupby`

<div style="max-width: 640px;">

The [`groupby`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) method allows you to group a DataFrame by a column and then calculate a sum, or any other statistic, for each unique value. This functions much like the <a href="https://en.wikipedia.org/wiki/Pivot_table">"pivot table"</a> feature found in most spreadsheets.

Let's use it to total up the accidents by make and model. You start by passing the field you want to group on to the function.

</div>

In [55]:
accident_list.groupby("latimes_make_and_model")

<div style="max-width: 640px;">

A nice start but you’ll notice you don’t get much back. The data’s been grouped, but we haven’t chosen what to do with it yet. If we wanted the total by model, we would use the `size` method.

</div>

In [57]:
accident_list.groupby("latimes_make_and_model").size()

latimes_make_and_model
AGUSTA 109                2
AIRBUS 130                1
AIRBUS 135                4
AIRBUS 350               29
BELL 206                 30
BELL 407                 13
HUGHES 369               13
MCDONNELL DOUGLAS 369     6
ROBINSON R22             20
ROBINSON R44             38
SCHWEIZER 269             5
SIKORSKY 76               2
dtype: int64

<div style="max-width: 640px;">
The result is much like `value_counts`, but we're allowed run to all kinds of statistical operations on the group, like `sum`, `mean` and `std`. For instance, we could sum the total number of fatalities for each maker by string that field on the end followed by the statistical method.
</div>

In [58]:
accident_list.groupby("latimes_make_and_model").total_fatalities.sum()

latimes_make_and_model
AGUSTA 109                5.0
AIRBUS 130                1.0
AIRBUS 135               11.0
AIRBUS 350               81.0
BELL 206                 61.0
BELL 407                 35.0
HUGHES 369               19.0
MCDONNELL DOUGLAS 369     7.0
ROBINSON R22             27.0
ROBINSON R44             71.0
SCHWEIZER 269             7.0
SIKORSKY 76              11.0
Name: total_fatalities, dtype: float64

<div style="max-width: 640px;">
Again our data has come back as an ugly Series. To reformat it as a pretty DataFrame use the `reset_index` method again.
</div>

In [59]:
accident_list.groupby("latimes_make_and_model").size().reset_index()

latimes_make_and_model   0
0              AGUSTA 109   2
1              AIRBUS 130   1
2              AIRBUS 135   4
3              AIRBUS 350  29
4                BELL 206  30
5                BELL 407  13
6              HUGHES 369  13
7   MCDONNELL DOUGLAS 369   6
8            ROBINSON R22  20
9            ROBINSON R44  38
10          SCHWEIZER 269   5
11            SIKORSKY 76   2

Now save that as a variable.

In [60]:
accident_counts = accident_list.groupby("latimes_make_and_model").size().reset_index()

<div style="max-width: 640px;">

You can clean up the `0` column name assigned by pandas with the `rename` method. The `inplace` option, found on many pandas methods, will save the change to your variable automatically.

</div>

In [62]:
accident_counts.rename(columns={0: "accidents"}, inplace=True)

The result is a DataFrame with the accident totals we'll want to merge with the FAA survey data to calculate rates.

In [64]:
accident_counts.head()

latimes_make_and_model  accidents
0             AGUSTA 109          2
1             AIRBUS 130          1
2             AIRBUS 135          4
3             AIRBUS 350         29
4               BELL 206         30

## Merging two dataframes together

<div style="max-width: 640px;">

Next we'll cover how to merge two DataFrames together into a combined table. Before we can do that, we need to read in a second file. We'll pull `faa-survey.csv`, which contains annual estimates of how many hours each type of helicopter was in the air. It was acquired via a Freedom of Information Act request with the FAA.

We can rip it in the same was the NTSB accident list, with `read_csv`.
    
</div>

In [78]:
survey = pd.read_csv("faa-survey.csv")

<div style="max-width: 640px;">

When joining two tables together, the first step is to look carefully at the columns in each table to find a common column that can be joined. We can do that with the `info` command we learned earlier.

</div>

In [80]:
accident_counts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 2 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   latimes_make_and_model  12 non-null     object
 1   accidents               12 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 320.0+ bytes


In [81]:
survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 2 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   latimes_make_and_model  12 non-null     object
 1   total_hours             12 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 320.0+ bytes


<div style="max-width: 640px;">

You can see that each table contains the `latimes_make_and_model` column. We can therefore join the two files using the pandas <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html">`merge`</a> method.

<div class="alert alert-block alert-warning"><p>If you are familar with traditional databases, you may recognize that the merge method in pandas is similar to SQL’s JOIN statement. If you dig into <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html">merge’s documentation</a> you will see it has many of the same options.</p></div>

Merging two DataFrames is as simple as passing both to pandas built-in `merge` method and specifying which field we’d like to use to connect them together. We will save the result into another new variable, which I'm going to call `merged_list`.
    
</div>

In [82]:
merged_list = pd.merge(accident_counts, survey, on="latimes_make_and_model")

That new DataFrame can be inspected like any other.

In [83]:
merged_list.head()

latimes_make_and_model  accidents  total_hours
0             AGUSTA 109          2       362172
1             AIRBUS 130          1      1053786
2             AIRBUS 135          4       884596
3             AIRBUS 350         29      3883490
4               BELL 206         30      5501308

By looking at the columns you can check how many rows survived the merge.

In [84]:
merged_list.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12 entries, 0 to 11
Data columns (total 3 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   latimes_make_and_model  12 non-null     object
 1   accidents               12 non-null     int64 
 2   total_hours             12 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 384.0+ bytes


<div style="max-width: 640px;">

You can also see that the dataframe now contains the same number of records as `accident_totals`. That's good. It means that every record in each dataframe found a match in the other. It's good idea to do a check like this every time you merge.

</div>

## Computing new columns

<div style="max-width:640px;">

Here's how you can create a new column based on the data in other columns, a process sometimes known as “computing.” In this case, computing can help us calculate a rate by dividing the accidents by flight hours.

In many cases, it's no more complicated than combining two series with a mathematical operator.
    
</div>

In [92]:
merged_list.accidents / merged_list.total_hours

0     5.522238e-06
1     9.489593e-07
2     4.521838e-06
3     7.467510e-06
4     5.453249e-06
5     6.150096e-06
6     1.081812e-05
7     1.089544e-05
8     6.732180e-06
9     1.610354e-05
10    4.388560e-06
11    2.184563e-06
dtype: float64

The resulting series can be added to your dataframe by assigning it to a new column.

In [93]:
merged_list['per_hour'] = merged_list.accidents / merged_list.total_hours

In [94]:
merged_list.head()

latimes_make_and_model  accidents  total_hours      per_hour
0             AGUSTA 109          2       362172  5.522238e-06
1             AIRBUS 130          1      1053786  9.489593e-07
2             AIRBUS 135          4       884596  4.521838e-06
3             AIRBUS 350         29      3883490  7.467510e-06
4               BELL 206         30      5501308  5.453249e-06

<div style="max-width:640px;">

In this case, the result is in scientific notation. As is common when calculating per capita statistics, you can multiple all results by a common number to make the numbers more legible.
    
That's as easy as tacking on the multiplication at the end of a computation.

</div>

In [95]:
merged_list['per_100k_hours'] = (merged_list.accidents / merged_list.total_hours) * 100000

## Sorting dataframes

<div style="max-width:640px;">

Another simple but common technique for analyzing data is sorting. This can be useful for ranking the DataFrame to show the highest and lowest members of the group according to a particular column. The <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html">`sort_values`</a> is how pandas does it.

</div>

In [98]:
merged_list.sort_values("per_100k_hours")

latimes_make_and_model  accidents  total_hours      per_hour  \
1              AIRBUS 130          1      1053786  9.489593e-07   
11            SIKORSKY 76          2       915515  2.184563e-06   
10          SCHWEIZER 269          5      1139326  4.388560e-06   
2              AIRBUS 135          4       884596  4.521838e-06   
4                BELL 206         30      5501308  5.453249e-06   
0              AGUSTA 109          2       362172  5.522238e-06   
5                BELL 407         13      2113788  6.150096e-06   
8            ROBINSON R22         20      2970806  6.732180e-06   
3              AIRBUS 350         29      3883490  7.467510e-06   
6              HUGHES 369         13      1201688  1.081812e-05   
7   MCDONNELL DOUGLAS 369          6       550689  1.089544e-05   
9            ROBINSON R44         38      2359729  1.610354e-05   

    per_100k_hours  
1         0.094896  
11        0.218456  
10        0.438856  
2         0.452184  
4         0.545325  
0         0.552224  
5         0.615010  
8         0.673218  
3         0.746751  
6         1.081812  
7         1.089544  
9         1.610354

<div style="max-width:640px;">

Note that returns the DataFrame resorted in ascending order from lowest to highest. That is pandas default way of sorting. Here's how you reverse it to show the largest values first.

</div>

In [99]:
merged_list.sort_values("per_100k_hours", ascending=False)

latimes_make_and_model  accidents  total_hours      per_hour  \
9            ROBINSON R44         38      2359729  1.610354e-05   
7   MCDONNELL DOUGLAS 369          6       550689  1.089544e-05   
6              HUGHES 369         13      1201688  1.081812e-05   
3              AIRBUS 350         29      3883490  7.467510e-06   
8            ROBINSON R22         20      2970806  6.732180e-06   
5                BELL 407         13      2113788  6.150096e-06   
0              AGUSTA 109          2       362172  5.522238e-06   
4                BELL 206         30      5501308  5.453249e-06   
2              AIRBUS 135          4       884596  4.521838e-06   
10          SCHWEIZER 269          5      1139326  4.388560e-06   
11            SIKORSKY 76          2       915515  2.184563e-06   
1              AIRBUS 130          1      1053786  9.489593e-07   

    per_100k_hours  
9         1.610354  
7         1.089544  
6         1.081812  
3         0.746751  
8         0.673218  
5         0.615010  
0         0.552224  
4         0.545325  
2         0.452184  
10        0.438856  
11        0.218456  
1         0.094896

## Further reading

<div style="max-width:640px;">

Congratulations. With that, we've recreated the analysis published in the Los Angeles Times and covered most of the basic skills necessary to access and analyze data with pandas. If you'd like to learn more, consult <a href="https://palewi.re/docs/first-python-notebook/">the full edition of "First Python Notebook"</a>.

</div>